[Merged by Bors] - feat(data/finsupp/basic): Version of `finsupp.prod_add_index` with weaker premises

stuart-presnell opened the pull request:
A simpler proof of `finsupp.prod_add_index : (f + g).prod h = f.prod h * g.prod h` with weaker premises.

Specifically, this only requires:
* `[add_zero_class M]` rather than `[add_comm_monoid M]`
* `h_zero : ∀ a ∈ f.support ∪ g.support, h a 0 = 1` rather than `h_zero : ∀a, h a 0 = 1`.  
* `h_add : ∀ (a ∈ f.support ∪ g.support) b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂` rather than `h_add : ∀ a b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂` (thanks to Anne Baanen for spotting this weakening).

The original version has been retained and re-named to `finsupp.prod_add_index'`, since in some places this is more convenient to use.  (There was already a lemma called `prod_add_index'` which appears not to have been used anywhere.  This has been renamed to `prod_hom_add_index`.)

Discussed in this Zulip thread:
https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Variant.20of.20finsupp.2Eprod_add_index.3F

---

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

stuart-presnell committed:
Add `prod_add_index''`

stuart-presnell committed:
Experimenting with replacing `prod_add_index` (fails)

Making this substitution causes `@[to_additive]` to complain for reasons I don't understand.

eric-wieser commented on src/data/finsupp/basic.lean:
```diff
@@ -1201,6 +1201,18 @@ lemma prod_add_index' [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
+      ←finset.prod_mul_distrib,
+      finsupp.prod_of_support_subset (f + g) finsupp.support_add h h_zero],
+  exact finset.prod_congr rfl (λ x hx, (by apply h_add)),
+end
```
It looks like this statement is more general than `prod_add_index` - can prove `prod_add_index` in terms of it? I'm thinking of something like:
```lean
-- this is the current proof of `prod_add_index`, with `λ a ha, h_zero a` replaced with `h_zero`
lemma prod_add_index'' [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
  {h : α → M → N} (h_zero : ∀ a ∈ f.support ∪ g.support, h a 0 = 1) (h_add : ∀a b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂) :
  (f + g).prod h = f.prod h * g.prod h :=
have hf : f.prod h = ∏ a in f.support ∪ g.support, h a (f a),
  from f.prod_of_support_subset (subset_union_left _ _) _ h_zero,
have hg : g.prod h = ∏ a in f.support ∪ g.support, h a (g a),
  from g.prod_of_support_subset (subset_union_right _ _) _ h_zero,
have hfg : (f + g).prod h = ∏ a in f.support ∪ g.support, h a ((f + g) a),
  from (f + g).prod_of_support_subset support_add _ h_zero,
by simp only [*, add_apply, prod_mul_distrib]

lemma prod_add_index [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
  {h : α → M → N} (h_zero : ∀a, h a 0 = 1) (h_add : ∀a b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂) :
  (f + g).prod h = f.prod h * g.prod h :=
prod_add_index'' (λ a ha, h_zero a) h_add
```
(untested)

stuart-presnell replied:
I suspect you're right.  I'm currently working on just replacing the original `prod_add_index` with this new variant, as Yaël [suggested](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/Variant.20of.20finsupp.2Eprod_add_index.3F/near/267461559).

eric-wieser replied:
Please link the Zulip discussion under the `---` in the PR description next time!

stuart-presnell replied:
Ah, sorry, I'll add that now!

stuart-presnell committed:
Add `@[to_additive]` to `prod_add_index''`

stuart-presnell committed:
Replace in `prod_finset_sum_index`

stuart-presnell committed:
Replace in `prod_option_index`

stuart-presnell committed:
Replacing in `prod_add_index'`

stuart-presnell committed:
Replacing in `prod_to_multiset`

stuart-presnell committed:
Replace in `eval₂_mul_monomial`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `sum_add_index` in `total_total`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Replace `finsupp.sum_add_index`

stuart-presnell committed:
Replace `sum_add_index`

stuart-presnell committed:
Golf proof of original `prod_add_index` via `prod_add_index''`

stuart-presnell committed:
Replace some `add_comm_monoid M` with `add_zero_class M`

stuart-presnell committed:
Merge branch 'master' into SP_prod_add_index

stuart-presnell committed:
Merge branch 'master' into SP_prod_add_index

stuart-presnell committed:
Merge branch 'SP_prod_add_index' of https://github.com/leanprover-community/mathlib into SP_prod_add_index

stuart-presnell committed:
Rename `*_add_index''` to `*_add_index` throughout

stuart-presnell committed:
Remove redundant line breaks

stuart-presnell committed:
Delete `prod_add_index_original`

stuart-presnell committed:
Fix `sum_update_add`

stuart-presnell committed:
Restore original `prod_add_index` as `prod_add_index'`

Also renames `prod_add_index'` and `sum_add_index'` as `...2`.  These lemmas don't appear to be used anywhere.

stuart-presnell committed:
Use `*_add_index'`

stuart-presnell committed:
Use `*_add_index'`

stuart-presnell committed:
Use `*_add_index'`

stuart-presnell committed:
Use `*_add_index'`

stuart-presnell committed:
Use `*_add_index'`

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
 
 @[simp]
-lemma sum_add_index' [add_comm_monoid M] [add_comm_monoid N] {f g : α →₀ M} (h : α → M →+ N) :
+lemma sum_add_index2 [add_zero_class M] [add_comm_monoid N] {f g : α →₀ M} (h : α → M →+ N) :
```
I'm not convinced by this name, how about something like:
```suggestion
lemma sum_hom_add_index [add_zero_class M] [add_comm_monoid N] {f g : α →₀ M} (h : α → M →+ N) :
```

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
 
 @[simp]
-lemma prod_add_index' [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
+lemma prod_add_index2 [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
```
```suggestion
lemma prod_hom_add_index [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
```

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
+      finsupp.prod_of_support_subset g (subset_union_right f.support _) h h_zero,
+      ←finset.prod_mul_distrib,
+      finsupp.prod_of_support_subset (f + g) finsupp.support_add h h_zero],
+  exact finset.prod_congr rfl (λ x hx, (by apply h_add)),
```
Does it make sense to also weaken to `h_add : ∀ (a ∈ f.support ∪ g.support) b₁ b₂, ...`?

stuart-presnell replied:
Ah yes, that works, thanks!  I've made that change and now I'm waiting to see what breaks.  :)

Vierkantor reviewed (approved):
Some small comments, looks good otherwise.

stuart-presnell committed:
Rename to `..._hom_add_index`

stuart-presnell committed:
Version with weaker `h_add` (and temp name)

stuart-presnell committed:
Delete version with just 1 weakened premise

stuart-presnell committed:
Fix in `prod_add_index'`

eric-wieser commented on src/data/finsupp/multiset.lean:
```diff
@@ -77,11 +77,9 @@ begin
-    { exact pow_zero },
-    { exact pow_add } }
+    rw [to_multiset_add, multiset.prod_add, ih, to_multiset_single, multiset.prod_nsmul,
+    finsupp.prod_add_index' pow_zero pow_add, finsupp.prod_single_index, multiset.prod_singleton],
```
```suggestion
      finsupp.prod_add_index' pow_zero pow_add, finsupp.prod_single_index, multiset.prod_singleton],
```

stuart-presnell committed:
Formatting

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
-lemma prod_add_index [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
+lemma prod_add_index [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
+  {h : α → M → N} (h_zero : ∀ a ∈ f.support ∪ g.support, h a 0 = 1)
+  (h_add  : ∀ (a ∈ f.support ∪ g.support) b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂) :
```
```suggestion
  (h_add : ∀ (a ∈ f.support ∪ g.support) b₁ b₂, h a (b₁ + b₂) = h a b₁ * h a b₂) :
```

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
 
 @[to_additive]
-lemma prod_add_index [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
+lemma prod_add_index [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
```
```suggestion
/-- Taking the product under `h` is an additive-to-multiplicative homomorphism of finsupps,
if `h` is an additive-to-multiplicative homomorphism on the support.

This is a more general version of `finsupp.prod_add_index'`; the latter has simpler hypotheses.
 -/
lemma prod_add_index [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
```

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
 finset.sum_sub_distrib
 
 @[to_additive]
```
```suggestion
@[to_additive "Taking the product under `h` is an additive homomorphism of finsupps,
if `h` is an additive homomorphism on the support.

This is a more general version of `finsupp.sum_add_index'`; the latter has simpler hypotheses."]
```

Vierkantor commented on src/data/finsupp/basic.lean:
```diff
@@ -1212,36 +1212,44 @@ lemma prod_inv [has_zero M] [comm_group G] {f : α →₀ M}
+end
+
+@[to_additive]
+lemma prod_add_index' [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
```
```suggestion
@[to_additive "Taking the product under `h` is an additive homomorphism of finsupps,
if `h` is an additive homomorphism.

This is a more specific version of `finsupp.sum_add_index` with simpler hypotheses."]
/-- Taking the product under `h` is an additive-to-multiplicative homomorphism of finsupps,
if `h` is an additive-to-multiplicative homomorphism.

This is a more specialized version of `finsupp.prod_add_index` with simpler hypotheses.
-/
lemma prod_add_index' [add_zero_class M] [comm_monoid N] {f g : α →₀ M}
```

Vierkantor reviewed (approved):
I'm happy after these changes. I'll let Oliver decide if the docstrings suffice or we also want to change the names.

bors d=@ocfnash

stuart-presnell committed:
Add docstrings per reviewer suggestion
